group-coordinator/src/test/java/org/apache/kafka/coordinator/group/streams/StreamsTopologyTest.java

```diff
@@ -0,0 +1,122 @@
+    private static final String CHANGELOG_TOPIC_1 = "changelog-1";
+    private static final String CHANGELOG_TOPIC_2 = "changelog-2";
+    private static final String CHANGELOG_TOPIC_3 = "changelog-3";
+
+    @Test
+    public void requiredTopicsShouldBeCorrect() {
+        Map<String, Subtopology> subtopologies = mkMap(
+            mkEntry(SUBTOPOLOGY_ID_1, mkSubtopology1()),
+            mkEntry(SUBTOPOLOGY_ID_2, mkSubtopology2())
+        );
+        StreamsTopology topology = new StreamsTopology(1, subtopologies);
+        Set<String> expectedTopics = Set.of(
+            SOURCE_TOPIC_1, SOURCE_TOPIC_2, SOURCE_TOPIC_3,
+            REPARTITION_TOPIC_1, REPARTITION_TOPIC_2, REPARTITION_TOPIC_3,
+            CHANGELOG_TOPIC_1, CHANGELOG_TOPIC_2, CHANGELOG_TOPIC_3
+        );
+
+        assertEquals(expectedTopics, topology.requiredTopics());
+    }
+
+    @Test
+    public void fromRecordShouldCreateCorrectTopology() {
+        StreamsGroupTopologyValue record = new StreamsGroupTopologyValue()
+            .setEpoch(1)
+            .setSubtopologies(Arrays.asList(mkSubtopology1(), mkSubtopology2()));
+        StreamsTopology topology = StreamsTopology.fromRecord(record);
+        assertEquals(1, topology.topologyEpoch());
+        assertEquals(2, topology.subtopologies().size());
+        assertTrue(topology.subtopologies().containsKey(SUBTOPOLOGY_ID_1));
+        assertEquals(mkSubtopology1(), topology.subtopologies().get(SUBTOPOLOGY_ID_1));
+        assertTrue(topology.subtopologies().containsKey(SUBTOPOLOGY_ID_2));
+        assertEquals(mkSubtopology2(), topology.subtopologies().get(SUBTOPOLOGY_ID_2));
+    }
+
+    private Subtopology mkSubtopology1() {
+        return new Subtopology()
+            .setSubtopologyId(SUBTOPOLOGY_ID_1)
+            .setSourceTopics(List.of(
+                SOURCE_TOPIC_1,
+                SOURCE_TOPIC_2,
+                REPARTITION_TOPIC_1,
+                REPARTITION_TOPIC_2
+            ))
+            .setRepartitionSourceTopics(List.of(
+                new TopicInfo().setName(REPARTITION_TOPIC_1),
+                new TopicInfo().setName(REPARTITION_TOPIC_2)
+            ))
+            .setRepartitionSinkTopics(List.of(
+                REPARTITION_TOPIC_3
+            ))
+            .setStateChangelogTopics(List.of(
+                new TopicInfo().setName(CHANGELOG_TOPIC_1),
+                new TopicInfo().setName(CHANGELOG_TOPIC_2)
+            ))
+            .setCopartitionGroups(List.of(
+                new StreamsGroupTopologyValue.CopartitionGroup()
+                    .setRepartitionSourceTopics(List.of((short) 0))
+                    .setSourceTopics(List.of((short) 0)),
+                new StreamsGroupTopologyValue.CopartitionGroup()
+                    .setRepartitionSourceTopics(List.of((short) 1))
+                    .setSourceTopics(List.of((short) 1))
+            ));
+    }
+
+    private Subtopology mkSubtopology2() {
```

Comment:
I personally preferred the unit tests when they were self-contained, which is easier to understand. There is no benefit to making this a separate method, and `mkSubtopology2` doesn't mean anything.

Reply:
I disagree. 
There is at least the benefit to not duplicate the code for the creation of the subtopology. 

What do you mean with
> `mkSubtopology2` doesn't mean anything

You look once at the `mkSubtopologyX` methods and understand the created subtopologies and then when you look at the tests you do not need to understand the subtopology again for each test.

Reply:
Okay, as stated this is just my preference - self-contained tests go a long way, and sharing data between tests easily becomes hard to deal with. This class won't grow much, so either way is probably fine. Agreed to keep it like this.